src/AMSlib/wf/basedb.hpp

```diff
@@ -1302,26 +1156,120 @@ class RMQConsumerHandler final : public AMQP::LibEventHandler
-  {
-    DBG(RMQConsumerHandler,
-        "[rank=%d] fatal error when establishing TCP connection: %s\n",
-        _rank,
-        message)
-  }
+                       const char* message) override;
 
   /**
-    *  Final method that is called. This signals that no further calls to your
+    *  @brief Final method that is called. This signals that no further calls to your
     *  handler will be made about the connection.
     *  @param  connection      The connection that can be destructed
     */
-  virtual void onDetached(AMQP::TcpConnection* connection) override
+  virtual void onDetached(AMQP::TcpConnection* connection) override;
+
+  bool waitFuture(std::future<RMQConnectionStatus>& future,
+                  unsigned ms,
+                  int repeat);
+};  // class RMQHandler
+
+/**
+ * @brief Specific handler for RabbitMQ connections based on libevent.
+ */
+class RMQConsumerHandler final : public RMQHandler
+{
+private:
+  /** @brief main channel used to send data to the broker */
+  std::shared_ptr<AMQP::TcpChannel> _channel;
+  /** @brief RabbitMQ queue (internal use only) */
+  std::string _queue;
+  /** @brief RabbitMQ exchange */
+  std::string _exchange;
+  /** @brief RabbitMQ routing key */
+  std::string _routing_key;
+  /** @brief Type of the exchange used (AMQP::topic, AMQP::fanout, AMQP::direct) */
+  AMQP::ExchangeType _extype;
+  /** @brief Queue that contains all the messages received on receiver queue */
+  std::shared_ptr<std::vector<AMSMessageInbound>> _messages;
+
+public:
+  /**
+   *  @brief Constructor
+   *  @param[in]  loop         Event Loop
+   *  @param[in]  cacert       SSL Cacert
+   *  @param[in]  routing_key  Routing key
+   *  @param[in]  exchange     Exchange
+   */
+  RMQConsumerHandler(uint64_t rId,
+                     std::shared_ptr<struct event_base> loop,
+                     std::string cacert,
+                     std::string exchange,
+                     std::string routing_key,
+                     AMQP::ExchangeType extype = AMQP::topic)
+      : RMQHandler(rId, loop, cacert),
+        _exchange(exchange),
+        _extype(extype),
+        _routing_key(routing_key),
+        _messages(std::make_shared<std::vector<AMSMessageInbound>>()),
+        _channel(nullptr)
   {
-    //  add your own implementation, like cleanup resources or exit the application
-    DBG(RMQConsumerHandler, "[rank=%d] Connection is detached.\n", _rank)
+    // Different queue name for each consumer (must be unique)
+    // In RMQ queues must be named excplicitly if we want to keep messages
+    // if unamed queue used, then messages sent before that queue is set up will be lost forever
+    // Important: queue names cannot be > 255 bytes
+    _queue = "amslib-" + std::to_string(rId);
+    CWARNING(RMQConsumerHandler,
+             _queue.length() > 255,
+             "Queue name \"%s\" is too long (%d)",
+             _queue.c_str(),
+             _queue.length())
```

Comment:
Can you add a NOTE/TODO comment. The fact that we expect to have a queue per rank may have performance implications.

Reply:
Done.